xarray/core/dataset.py

```diff
@@ -2337,21 +2337,36 @@ def expand_dims(self, dim, axis=None):
 
         Parameters
         ----------
-        dim : str or sequence of str.
+        dim : str, sequence of str, dict, None
             Dimensions to include on the new variable.
-            dimensions are inserted with length 1.
+            If provided as str or sequence of str, then dimensions are inserted
+            with length 1. If provided as a dict, then the keys are the new
+            dimensions and the values are either integers (giving the length of
+            the new dimensions) or iterables (giving the coordinates of the new
+            dimensions).
         axis : integer, list (or tuple) of integers, or None
             Axis position(s) where new axis is to be inserted (position(s) on
             the result array). If a list (or tuple) of integers is passed,
             multiple axes are inserted. In this case, dim arguments should be
             the same length list. If axis=None is passed, all the axes will
             be inserted to the start of the result array.
+        **new_dims : int or  iterable
+            The keywords are arbitrary dimensions being inserted and the values
+            are either the lengths of the new dims (if int is given), or their
+            coordinates. Note, this is an alternative to passing a dict to the
+            dim kwarg and will only be used if dim is None.
 
         Returns
         -------
         expanded : same type as caller
             This object, but with an additional dimension(s).
         """
+        if dim is None:
+            dim = new_dims
+        elif dim and new_dims:
+            raise ValueError('cannot specify both keyword and positional '
+                             'arguments to expand_dims')
```

Comment:
Probably it would be a little cleaner if we can make `dim` always a dict
```python
if dim is not None and not utils.is_dict_like(dim):
    dim = {dim: 1}
dim = either_dict_or_kwargs(dim, dim_kwargs, 'expand_dims')
```